Remove biprime context setup and merge static biprime into servkey context

srcansiz opened the pull request:
**PR description**

- Removes non-implemented biprime dynamic generation
- merges static biprime public param into serv/user key context
- Updates unit-tests
- Updates configuration file

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.



**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.


**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state_breakpoint`/`load_state_breakpoint` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`

mvesin commented:
To be merged after next release

mvesin commented on fedbiomed/researcher/secagg/_secagg_context.py:
```diff
@@ -21,8 +21,12 @@
 from fedbiomed.common.mpc_controller import MPCController
-from fedbiomed.common.secagg_manager import BaseSecaggManager, SecaggServkeyManager, SecaggBiprimeManager, SecaggDhManager
-from fedbiomed.common.utils import matching_parties_servkey, matching_parties_biprime, get_method_spec
+from fedbiomed.common.secagg_manager import BaseSecaggManager, SecaggServkeyManager, SecaggDhManager
```
Misc: can remove obsolete imports
```
import time
import random
```

mvesin commented on fedbiomed/researcher/secagg/_secagg_context.py:
```diff
@@ -158,11 +161,10 @@ def set_experiment_id(self, experiment_id: Union[str, None]) -> None:
-                      f'context is set for biprime.'
-            logger.error(errmess)
-            raise FedbiomedSecaggError(errmess)
+        if not isinstance(experiment_id, str):
```
Misc: please update accordingly the prototype and docstrings including for:
- SecaggContext.__init__()
- SecaggContext.experiment_id()
- SecaggContext.set_experiment_id()
- SecaggMpspdzContext.__init__()

mvesin replied:
`experiment_id` type `Union[str, None]` was existing specifically for the case of biprime. For server key and DH, they are always associated to one experiment thus `experiment_id` type should now always be `str`

We should also propagate that to:
- `fedbiomed.researcher.secure_aggregation` (this is only argument type/docstrings/class attribute type updates)

mvesin commented on fedbiomed/common/constants.py:
```diff
@@ -260,12 +248,10 @@ class SecaggElementTypes(_BaseEnum):
     SERVER_KEY: int = 0
-    BIPRIME: int = 1
-    DIFFIE_HELLMAN: int = 2
+    DIFFIE_HELLMAN: int = 1
```
As you renumbered the `SecaggElementTypes` you need to increase the major version of messaging protocol and secagg element

```
__messaging_protocol_version__ = FBM_Component_Version('5')  # format of gRPC messages.
__secagg_element_version__ = FBM_Component_Version('3')  # format os secagg database elements
```

With current code, a node before this PR would accept to negotiate secagg with a researcher with this PR (or vice versa), but would mistake a biprime secagg negotiation for a DH negotiation (or refuse to do a DH nego etc.)

srcansiz replied:
Hi @mvesin,

Thanks for catching it. Then, I am not going to change the element type value. It will avoid changing the protocol version.

mvesin commented on fedbiomed/common/config.py:
```diff
@@ -160,43 +147,36 @@ def generate(
 
         ip, port = retrieve_ip_and_port(self.root)
-        allow_default_biprimes = os.getenv('ALLOW_DEFAULT_BIPRIMES', True)
+        allow_default_biprimes = os.getenv("ALLOW_DEFAULT_BIPRIMES", True)
```
This line is not needed anymore

```suggestion
        allow_default_biprimes = os.getenv("ALLOW_DEFAULT_BIPRIMES", True)
```

mvesin commented on fedbiomed/common/secagg_manager.py:
```diff
@@ -13,8 +13,11 @@
 
 from fedbiomed.common.utils import raise_for_version_compatibility, __default_version__
-from fedbiomed.common.constants import ErrorNumbers, BiprimeType, SecaggElementTypes, \
+from fedbiomed.common.constants import (
```
Misc: please also remove now obsolete imports

```
import os
import json
from fedbiomed.common.validator import Validator, ValidatorError, SchemeValidator
```

mvesin commented on fedbiomed/node/secagg/_secagg_setups.py:
```diff
@@ -19,11 +19,12 @@
 from fedbiomed.common.mpc_controller import MPCController
 from fedbiomed.common.secagg import DHKey, DHKeyAgreement
 from fedbiomed.common.synchro import EventWaitExchange
+from fedbiomed.common.utils import get_default_biprime
```
Misc : please also remove obsoleted imports

```
import time
import random
```

mvesin commented on fedbiomed/common/secagg_manager.py:
```diff
@@ -13,8 +13,11 @@
+    ErrorNumbers,
+    SecaggElementTypes,
     __secagg_element_version__
+)
```
In all the `fedbiomed.common.secagg_manager` module, we should probably update `experiment_id` type from `Union[str, None]` to `str` (`None` was existing only for the biprime case).

This means updating argument types, docstrings, plus a little bit of code in `_raise_error_incompatible_requested_entry`

mvesin commented on fedbiomed/node/secagg/_secagg_setups.py:
```diff
@@ -19,11 +19,12 @@
 
 from fedbiomed.node.environ import environ
-from fedbiomed.node.secagg_manager import SKManager, BPrimeManager, DHManager, SecaggManager
+from fedbiomed.node.secagg_manager import SKManager, DHManager, SecaggManager
```
in this module also we need to update type of `experiment_id` from `Union[str, None]` to `str`

mvesin commented on fedbiomed/node/secagg/_secagg_setups.py:
```diff
@@ -238,11 +239,6 @@ def __init__(
         self._element = SecaggElementTypes.SERVER_KEY
         self._secagg_manager = SKManager
 
-        if not self._experiment_id or not isinstance(self._experiment_id, str):
```
I suspect removing this test introduces a security hole (or room for a potential security hole). We need to ensure that a secagg context is attached to one experiment and cannot be used by another experiment. Thus we need to ensure the experiment id is not the empty string (which would make guessing really easy ...).

Currently in `fedbiomed.common.message` we have

```
class SecaggRequest(RequestReply, RequiresProtocolVersion):
...
      experiment_id: Optional[str]
```

So we have no guarantee about the type and value of the data delivered to `SecaggServkeySetup.__init__()`

What we could do now to simplify the code would be:

1. change `fedbiomed.common.message` to

    ```
    class SecaggRequest(RequestReply, RequiresProtocolVersion):
    ...
      experiment_id: str
    ```
3. then replace this test with a test in the base class constructor `fedbiomed.node.secagg.SecaggBaseSetup.__init__()` as we now know `experiment_id` is a `str`, and we just need to check it is non null

    ```
    if not self._experiment_id:
            errmess = f'{ErrorNumbers.FB318.value}: bad parameter `experiment_id` must be a non empty string'
            logger.error(errmess)
            raise FedbiomedSecaggError(errmess)
    ```

4. remove the equivalent test in `SecaggDHSetup.__init__()` as it is replaced by the simplified test in the base class in step 3.

mvesin replied:
As discussed: changing `experiment_id` on `SecaggRequest` message is enough.
No need to ensure uniqueness (and secretness) of `experiment_id` currently. There might be need to partition data in the future when multiple researchers will be implemented, but the way of doing that will be decided later (and probably more scure/general mechanism)

mvesin commented on docs/user-guide/nodes/configuring-nodes.md:
```diff
@@ -87,8 +85,6 @@ private_key = certs/cert_node_73768d5f-6a66-47de-8533-1291c4ef59d1/MPSPDZ_certif
 public_key = certs/cert_node_73768d5f-6a66-47de-8533-1291c4ef59d1/MPSPDZ_certificate.pem
 mpspdz_ip = localhost
 mpspdz_port = 14004
-allow_default_biprimes = True
```
Thank you for searching the doc for obsoleted references to biprime. I found a few more to be updated
- docs/tutorials/security/secure-aggregation.ipynb references `exp.secagg.biprime` which does not exist anymore
- docs/user-guideresearcher/experiment.md details in json format the content of the train request listing a `secagg_biprime_id` which does not exist (this doc was already not up to date as the 4 secagg_* were replaced with `secagg_arguments` already )
- docs/user-guide/secagg/introduction.md explains generation of biprime which was removed
- docs/user-guide/secagg/researcher-interface.md mentions generation of biprime
- tests/test_message.py has a test with a secagg_biprime_id argument (which is correct for a unit test but might now be confusing, we can choose to remove this line)

srcansiz replied:
Thank you, I updated those appearances.

mvesin commented:
Thank you @srcansiz for this nice work !

I now completed the review + tested the notebook

The 2 important points in the review are:
- needed update of versioning
- potential secuiry issue due to the removal of checking of `experiment_id` in node secagg setup

Others are mostly polishing (doc, typing of experiment_id).

Once they are adressed, ok for me to merge !

srcansiz commented:
> Thank you @srcansiz for this nice work !
> 
> I now completed the review + tested the notebook
> 
> The 2 important points in the review are:
> 
> * needed update of versioning
> * potential secuiry issue due to the removal of checking of `experiment_id` in node secagg setup
> 
> Others are mostly polishing (doc, typing of experiment_id).
> 
> Once they are adressed, ok for me to merge !

Hi @mvesin,

Thank you for the review and your comments. I've implemented all your suggestions. 
- For versioning: I revert back to Deffie-Helman key element to `2` to keep it compatible. 
- I converted `experiment_id` try in secagg request from optional to str.

mvesin commented:
Thank you @srcansiz !
Pushed a few final glitches => OK for me for merging
